src/eventHandlers/InvocationHandler.ts

```diff
@@ -90,16 +90,21 @@ export class InvocationHandler extends EventHandler<'invocationRequest', 'invoca
                 });
             });
 
-            const hookData: HookData = {};
             let userFunction = channel.functionLoader.getFunc(functionId);
+
+            // create a copy of the hook data in the worker context (set by app hooks)

```

Comment:
I like the idea of providing the app-level data to invocation-level hooks, but I'm not so sure about copying like this. For example, the `...` operator is just a shallow copy (let me know if you need me to explain "shallow copy"), which may lead to unexpected behavior.

What if we add a property `appHookData` to `PreInvocationContext` and `PostInvocationContext`. Aka we don't do any copying and it's just a separate object. We could warn users that changes to the data will be app-wide and it's up to them if they want to create a copy

Reply:
Oh I didn't know `...` only did a shallow copy, thank you for pointing that out. We could do a deep copy to get around that issue.

Alternatively, I do like the idea of a separate `appHookData` for invocation contexts. It also has the added benefit of allowing different invocations to speak to each other, which is dangerous but could be useful. In that case, would it still be the `hookData` property for `appStartupContext`? Or do we still want to keep some data property that is only shared by app-level hooks and not accessible by individual invocations?

Reply:
I think appStartupContext should have an `appHookData` property that is the one shared with invocation hook contexts. I'm just not sure if appStartupContext should _also_ have a `hookData` property for just app-level hooks. It might be a little confusing, but I guess it can't hurt to have both

Reply:
Yeah I agree it can be very confusing if we have both. We might need to think about how to name them to make sure the distinction is clear.

Reply:
Thinking about it, what if we make the `hookData` property the one that is shared between all hooks, including app startup and different invocations, and we add a new property `invocationHookData` that is tied to one invocation?

Reply:
Done